packages/core/src/compiler.ts

```diff
@@ -303,21 +309,28 @@ export default class Compiler {
     }
   }
 
-  setFormatter(key: string) {
+  setFormatter(key: string, parentKey?: string) {
     if (this.runtimeIncludes(key, 'formatter')) return;
-    let cf = this.options.customFormatters[key];
+    const cf = this.options.customFormatters[parentKey || key];
     if (cf) {
-      if (typeof cf === 'function') cf = { formatter: cf };
+      const cfo = typeof cf === 'function' ? { formatter: cf } : cf;
+
       this.runtime[key] = Object.assign(
-        cf.formatter,
+        cloneFunction(cfo.formatter),
         { type: 'formatter' } as const,
         'module' in cf && cf.module && cf.id
-          ? { id: identifier(cf.id), module: cf.module }
+          ? {
+              id: identifier(cf.id),
+              module:
+                typeof cf.module === 'function' && this.plural.locale
+                  ? cf.module({ locale: this.plural.locale })
+                  : cf.module
+            }
           : { id: null, module: null }
       );
     } else if (isFormatterKey(key)) {
       this.runtime[key] = Object.assign(
-        Formatters[key],
+        cloneFunction(Formatters[key]),
```

Comment:
OTOH, here the formatter is an internal one, so we don't need to clone it and can mutate it without ill effect.

Reply:
it's your call. i am biased to _not mutating the underlying fn_ ever, out of what I perceive to be general purpose best practices, but that is entirely subjective :)

Reply:
Let's just mutate it, and inline the `cloneFunction()` above.

Reply:
done, 9ee2bf1